django/db/models/sql/compiler.py

```diff
@@ -223,7 +223,12 @@ def get_select_from_parent(klass_info):
 
         ret = []
         for col, alias in select:
-            ret.append((col, self.compile(col, select_format=True), alias))
+            try:
+                sql, params = self.compile(col, select_format=True)
+            except EmptyResultSet:
+                # Selecting a predicate that is always false:
+                sql, params = '0', ()
```

Comment:
This isn't safe unfortunately. Consider the following test results (which fail):

```
    def test_empty_expression_char(self):
        books = Book.objects.annotate(
            selected=Case(
                When(pk__in=Book.objects.none(), then=Value('Empty')),
                default=Value('Not Empty'), output_field=CharField()
            )
        )
        self.assertGreater(len(books), 0)
        self.assertEqual(books[0].selected, 'Not Empty')

    def test_empty_expression_datetime(self):
        from django.utils import timezone
        from datetime import timedelta
        now = timezone.now()
        then = now + timedelta(days=1)

        books = Book.objects.annotate(
            selected=Case(
                When(pk__in=Book.objects.none(), then=Value(now)),
                default=Value(then), output_field=DateTimeField()
            )
        )
        self.assertGreater(len(books), 0)
        self.assertEqual(books[0].selected, then)
```

Depending on whether or not there are dbconverters run, these will fail with one of the following kinds of errors:

1. `TypeError: expected string or buffer`
2. `AssertionError: 0 != datetime.datetime(2016, 5, 23, 18, 59, 26, 409777)`

For what it's worth, both of these tests fail without your patch too, except they fail during the count just like the report on the ticket says. This certainly needs fixing in some way, but we can't just use 0 as a value all of the time.